### PR TITLE
Add time derivative computation to ScalarTensor

### PR DESCRIPTION
**Pull request by guilara**
## Proposed changes
We compute the time derivative of the combined ScalarTensor system by calling the computation from each of the individual (CSW and GH) systems. Additionally, we add any sources to the RHS of the equations to include the back-reaction of the spacetime and source terms in the wave equation of the scalar.

In the current approach, the geometrical quantities needed by CSW are provided by compute tags. An optimization of this might be to add them as temporaries (or to store them in cache?). However, we may do this in another PR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+#include "Evolution/Systems/CurvedScalarWave/System.hpp"
+#include "Evolution/Systems/CurvedScalarWave/TimeDerivative.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/System.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/TimeDerivative.hpp"
+#include "Evolution/Systems/ScalarTensor/Sources/ScalarSource.hpp"
+#include "Evolution/Systems/ScalarTensor/StressEnergy.hpp"
+#include "Evolution/Systems/ScalarTensor/System.hpp"
+#include "Evolution/Systems/ScalarTensor/Tags.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+// Tag obtained from gh::TimeDerivative needs to be complete?
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: whats this?

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+ * `CurvedScalar::TimeDerivativeTerms` as a 'product system' -- each
+ * independently operating on its own subset of the supplied variable
+ * collections.
+ * The additional steps taken are to compute the trace-reversed stress energy
+ * tensor associated with the scalar part of the system and add its contribution
+ * to the \f$\partial_t \Pi_{a b}\f$ variable in the Generalized Harmonic
+ * system, as well as adding any scalar sources to the variable \f$\partial_t
+ * \Pi\f$.
+ */
+struct TimeDerivativeTerms {
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Just call this `TimeDerivative` to make it consistent with the other systems. Same for the filename

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+#include "Time/Tags/Time.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace ScalarTensor {
+/*!
+ * \brief Compute the RHS terms and flux values for both the Generalized
+ * Harmonic formulation of Einstein's equations and the scalar equations.
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: This sounds like there is no coupling at all. how about "... for a combined system of the Generalized ..."

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+ * The additional steps taken are to compute the trace-reversed stress energy
+ * tensor associated with the scalar part of the system and add its contribution
+ * to the \f$\partial_t \Pi_{a b}\f$ variable in the Generalized Harmonic
+ * system, as well as adding any scalar sources to the variable \f$\partial_t
+ * \Pi\f$.
+ */
+struct TimeDerivativeTerms {
+  using gh_dt_tags =
+      db::wrap_tags_in<::Tags::dt,
+                       typename gh::System<3_st>::variables_tag::tags_list>;
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: I would template this class on Dim, or create a `static constexpr size_t Dim = 3` at the top to make it consistent with the other systems.

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  using scalar_dt_tags = db::wrap_tags_in<
+      ::Tags::dt,
+      typename CurvedScalarWave::System<3_st>::variables_tag::tags_list>;
+  using dt_tags = tmpl::append<gh_dt_tags, scalar_dt_tags>;
+  using scalar_flux_tags = tmpl::transform<
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: you are not using this anywhere outside tests. from what I can tell. Also, `CSW` does not have any flux_variables, so I don't understand what you are trying to do here.

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      const tnsr::I<DataVector, 3_st>& shift_scalar,
+      const tnsr::i<DataVector, 3_st>& deriv_lapse,
+      const tnsr::iJ<DataVector, 3_st>& deriv_shift,
+      const tnsr::II<DataVector, 3_st>& upper_spatial_metric,
+      const tnsr::I<DataVector, 3_st>& trace_spatial_christoffel,
+      const Scalar<DataVector>& trace_extrinsic_curvature,
+      const Scalar<DataVector>& gamma1_scalar,
+      const Scalar<DataVector>& gamma2_scalar,
+
+      const Scalar<DataVector>& scalar_source) {
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: move the definition to a .cpp file

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <utility>
```
Comment (nikwit) [inline src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: are you using this?

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/IsA.hpp"
+
+namespace {
+template <typename ComputeVolumeTimeDerivativeTerms, size_t Dim,
+          typename EvolvedTagList, typename FluxTagList, typename TempTagList,
+          typename GradientTagList, typename ArgTagList>
+struct ComputeVolumeTimeDerivativeTermsHelper;
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: can you please just call the three systems separately without a helper that is very hard to read? Tests in particular should always be very readable and it doesn't matter if they are a bit longer.

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+      [[maybe_unused]] const gsl::not_null<FluxVariables*> volume_fluxes,
+      const gsl::not_null<TemporaryVariables*> temporaries,
+      const GradientVariables& partial_derivs,
+      const ArgumentVariables& time_derivative_args) {
+    ComputeVolumeTimeDerivativeTerms::apply(
+        make_not_null(&get<::Tags::dt<EvolvedTags>>(*dt_vars_ptr))...,
+        make_not_null(&get<FluxTags>(*volume_fluxes))...,
+        make_not_null(&get<TempTags>(*temporaries))...,
+        get<GradientTags>(partial_derivs)..., [](const auto& t) -> const auto& {
+          if constexpr (tt::is_a_v<std::unique_ptr,
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: do the systems have any arg tags that are unique pointers?

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  using flux_variables_type =
+      Variables<tmpl::append<gh_flux_tags, scalar_flux_tags>>;
+
+  using gh_temp_tags = ScalarTensor::TimeDerivativeTerms::gh_temp_tags;
+  using scalar_temp_tags =
+      ScalarTensor::TimeDerivativeTerms::scalar_temp_tags;
+  using temp_variables_type = Variables<
+      typename ScalarTensor::TimeDerivativeTerms::temporary_tags>;
+
+  using gh_gradient_tags = tmpl::transform<
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: doesn't `using gh_gradient_tags = ScalarTensor::TimeDerivativeTerms::gh_gradient_tags` work fine?

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+  using gh_arg_tags =
+      ScalarTensor::TimeDerivativeTerms::gh_arg_tags;
+  using scalar_arg_tags =
+      ScalarTensor::TimeDerivativeTerms::scalar_arg_tags;
+  using arg_variables_type = tuples::tagged_tuple_from_typelist<
+      tmpl::append<gh_arg_tags, scalar_arg_tags,
+                   tmpl::list<ScalarTensor::Tags::ScalarSource>>>;
+
+  const size_t element_size = 10_st;
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: 10 works fine

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+  temp_variables_type expected_temp_variables{element_size};
+  temp_variables_type temp_variables{element_size};
+
+  const auto gradient_variables =
+      make_with_random_values<gradient_variables_type>(
+          make_not_null(&gen), make_not_null(&dist), DataVector{element_size});
+
+  arg_variables_type arg_variables;
+  tmpl::for_each<tmpl::append<
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: It' would be a lot more readable to just remove the `MeshVelocity` from the tag list and create it separately so you don't need all of this.

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+      typename flux_variables_type::tags_list,
+      typename temp_variables_type::tags_list,
+      typename gradient_variables_type::tags_list,
+      typename arg_variables_type::tags_list>::apply(
+                              make_not_null(&dt_variables),
+                              make_not_null(&flux_variables),
+                              make_not_null(&temp_variables),
+                              gradient_variables, arg_variables);
+
+  // We compute the trace-reversed stress energy tensor fot the expected
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: typo fot

**Review by nikwit — CHANGES_REQUESTED** [review]
nice work, here are some suggestions

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 9 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: Apparently `gh::gauges::GaugeCondition` has a unique pointer

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 1 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Without the include below this comment I get some error in the test that Time is an incomplete tag. Don't really understand this. It arises from something recently merged in develop it seems.

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 2 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Changed name

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 3 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Changed description to just say we compute the rhs of ScalarTensor

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 4 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Added the `static constexp size_t Dim` param in the declaration, But in the .cpp file I still need to write `3_st` or sth ugly like `TimeDerivative::Dim` (I think), so left it with `3_st` in the .cpp file

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 5 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Yeah, my implementation of all this is heavily based in GhMhd so thats why there are some of these seemegly strange and useless params sometimes. I removed all the flux arguments from there and from the test

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 6 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Moved

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 7 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Seems not. Removed `utility`

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 8 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: As above, I was trying to reuse as much as possible from GhMhd. This was coded up for the `TimeDerivative` test there and it seemed useful. I removed a few things from it that were not really used. 

... I guess I could try to do without it....

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 10 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: Tried it, doesn't seem to work for the Helper

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 11 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: Just 10 :)

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 12 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: Same comment as for the Helper. I guess I could also try to dismantle this...

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 13 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: Fixed

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivative.cpp [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+#include "PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/DerivSpatialMetric.hpp"
+#include "PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/ExtrinsicCurvature.hpp"
+#include "PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfLapse.hpp"
+#include "PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfShift.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/IsA.hpp"
+
+#include "Parallel/Printf.hpp"
```
Comment (nikwit) [inline tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivative.cpp]: remove

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 1 above)
Comment (nikwit) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: ok, I think you can remove the comment.

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 4 above)
Comment (nikwit) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: I think you can just write Dim in the cpp file.

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 8 above)
Comment (nikwit) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: I think the test would be a lot more readable if you called them separately. It might be a bit wordy but should be quick to write... and this way you don't have to wrap your head around these helper and metafunctions when reading a simple test. Just call them similar as its done`CurvedScalarWave/TestTimeDerivative:104`

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp (hunk 10 above)
Comment (nikwit) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivativeTerms.cpp]: ah right, I misread this. The gradient tags in the system don't actually have the derivative tag attached to it yet and this metafunction does that. fine as is

**Review by nikwit — CHANGES_REQUESTED** [review]
Looks good, I think changing the test would be nice. Also, please do not squash fixup changes directly into your commits unless told otherwise by the reviewer.

**Comment by guilara** [comment]
> Looks good, I think changing the test would be nice. Also, please do not squash fixup changes directly into your commits unless told otherwise by the reviewer.

Ah, sorry. That makes sense

File: tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivative.cpp (hunk 14 above)
Comment (guilara) [reply tests/Unit/Evolution/Systems/ScalarTensor/Test_TimeDerivative.cpp]: Removed

File: src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp (hunk 4 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivativeTerms.hpp]: Added. That worked

**Comment by guilara** [comment]
I've rewritten the test and removed the Helper at all. It's a bit lengthy.

**Comment by guilara** [comment]
I also removed the `_st`s

**Review by nikwit — APPROVED** [review]
Looks good, you can squash the commits together :)

File: src/Evolution/Systems/ScalarTensor/TimeDerivative.cpp [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Time/Tags/Time.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace ScalarTensor {
+void TimeDerivative::apply(
+    // GH dt variables
+    gsl::not_null<tnsr::aa<DataVector, Dim>*> dt_spacetime_metric,
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarTensor/TimeDerivative.cpp]: `gsl::` -> `const gsl::` (find-replace in the cpp will be easiest probably)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
LGTM, one minor suggestion. I think you can just squash all the commits into 1, right? I don't need to see a fixup for the suggestion I'm making.

File: src/Evolution/Systems/ScalarTensor/TimeDerivative.cpp (hunk 15 above)
Comment (guilara) [reply src/Evolution/Systems/ScalarTensor/TimeDerivative.cpp]: I've changed it. Also squashed the commits

**Comment by guilara** [comment]
I've also changed `Dim` to `dim` in `TimeDerivative` since its not a template parameter and to be consistent with #5235

**Review by nilsdeppe — APPROVED** [review]
